scene/gui/code_edit.cpp

```diff
@@ -755,7 +769,34 @@ void CodeEdit::_handle_unicode_input_internal(const uint32_t p_unicode, int p_ca
 				} else if (cc < get_line(cl).length() && !is_symbol(get_line(cl)[cc])) {
 					insert_text_at_caret(chr, i);
 				} else if (post_brace_pair != -1 && auto_brace_completion_pairs[post_brace_pair].close_key[0] == chr[0]) {
-					caret_move_offset = auto_brace_completion_pairs[post_brace_pair].close_key.length();
+					// Try to count the brace balance for this pair. Insert closing brace if there are too much open braces, otherwise just move the cursor.
+					int pair_balance = 0;
+					if (auto_brace_completion_pairs[post_brace_pair].close_key.length() == 1 && auto_brace_completion_pairs[post_brace_pair].open_key.length() == 1) {
+						// HACK: We only want to balance in the same scope (or even expression). But scope information is not exposed by the script language so indentation is our best heuristic.
+						int indent_level = get_indent_level(cl);
+
+						for (int balance_line = cl; balance_line >= 0; balance_line--) {
+							if (get_indent_level(balance_line) < indent_level) {
+								// Less indented -> not part of the current scope -> stop searching
+								break;
```

Comment:
Do we have anything against inline comments for breaks?
```suggestion
								break; // Less indented -> not part of the current scope -> stop searching.
```